OpenRA.Mods.Common/Traits/World/DomainIndex.cs

```diff
@@ -9,246 +9,38 @@
-			toProcess.Push(d1);
-
-			while (toProcess.Any())
-			{
-				var current = toProcess.Pop();
-				if (!transientConnections.ContainsKey(current))
-					continue;
-
-				foreach (var neighbor in transientConnections[current])
-				{
-					if (neighbor == d2)
-						return true;
-
-					if (!visited.Contains(neighbor))
-						toProcess.Push(neighbor);
-				}
-
-				visited.Add(current);
-			}
-
-			return false;
-		}
-
-		void CreateConnection(ushort d1, ushort d2)
-		{
-			if (!transientConnections.ContainsKey(d1))
-				transientConnections[d1] = new HashSet<ushort>();
-			if (!transientConnections.ContainsKey(d2))
-				transientConnections[d2] = new HashSet<ushort>();
-
-			transientConnections[d1].Add(d2);
-			transientConnections[d2].Add(d1);
-		}
+		readonly PathFinder pathFinder;
 
-		bool CanTraverseTile(World world, CPos p)
+		public DomainIndex(Actor self)
 		{
-			if (!map.Contains(p))
-				return false;
-
-			var terrainOffset = world.Map.GetTerrainIndex(p);
-			return (movementClass & (1 << terrainOffset)) > 0;
+			pathFinder = self.Trait<PathFinder>();
 		}
 
-		void BuildDomains(World world)
+		public bool IsPassable(CPos p1, CPos p2, Locomotor locomotor)
 		{
-			ushort domain = 1;
-
-			var visited = new CellLayer<bool>(map);
-
-			var toProcess = new Queue<CPos>();
-			toProcess.Enqueue(MPos.Zero.ToCPos(map));
-
-			// Flood-fill over each domain.
-			while (toProcess.Count != 0)
-			{
-				var start = toProcess.Dequeue();
-
-				// Technically redundant with the check in the inner loop, but prevents
-				// ballooning the domain counter.
-				if (visited[start])
-					continue;
-
-				var domainQueue = new Queue<CPos>();
-				domainQueue.Enqueue(start);
-
-				var currentPassable = CanTraverseTile(world, start);
-
-				// Add all contiguous cells to our domain, and make a note of
-				// any non-contiguous cells for future domains.
-				while (domainQueue.Count != 0)
-				{
-					var n = domainQueue.Dequeue();
-					if (visited[n])
-						continue;
-
-					var candidatePassable = CanTraverseTile(world, n);
-					if (candidatePassable != currentPassable)
-					{
-						toProcess.Enqueue(n);
-						continue;
-					}
-
-					visited[n] = true;
-					domains[n] = domain;
-
-					// PERF: Avoid LINQ.
-					foreach (var direction in CVec.Directions)
-					{
-						// Don't crawl off the map, or add already-visited cells.
-						var neighbor = direction + n;
-						if (visited.Contains(neighbor) && !visited[neighbor])
-							domainQueue.Enqueue(neighbor);
-					}
-				}
-
-				domain += 1;
-			}
-
-			Log.Write("debug", "Found {0} domains for movement class {1} on map {2}.", domain - 1, movementClass, map.Title);
+			// TODO: We should remove DomainIndex wholesale and instead expose
+			// PathExists on the IPathFinder interface.
+			// Otherwise we can never replace the concrete PathFinder with other implementations.
+			// Need to consider whether we need to expose Locomotor as a hard requirement in the interface,
```

Comment:
Thinking out loud: Would it make sense to invert this relationship, so that `Locomotor` provides the public interface and the pathfinder is an implementation detail of that?

Reply:
I considered your thought plus the original options I outlined in the comment. Exposing the `PathExists` method (and `Locomotor`) on the `IPathFinder` interface still felt like the best option in the end, so that's what I have done.